Lib/subprocess.py

```diff
@@ -1066,6 +1066,34 @@ def wait(self, timeout=None):
                 pass
             raise  # resume the KeyboardInterrupt
 
+    def _close_pipe_fds(self,
+                        p2cread, p2cwrite,
+                        c2pread, c2pwrite,
+                        errread, errwrite):
+        # self._devnull is not always defined.
+        devnull_fd = getattr(self, '_devnull', None)
+
+        if _mswindows:
+            if p2cread != -1:
+                p2cread.Close()
+            if c2pwrite != -1:
+                c2pwrite.Close()
+            if errwrite != -1:
+                errwrite.Close()
+        else:
+            if p2cread != -1 and p2cwrite != -1 and p2cread != devnull_fd:
+                os.close(p2cread)
+            if c2pwrite != -1 and c2pread != -1 and c2pwrite != devnull_fd:
+                os.close(c2pwrite)
+            if errwrite != -1 and errread != -1 and errwrite != devnull_fd:
+                os.close(errwrite)
+
+        if devnull_fd is not None:
+            os.close(devnull_fd)
```

Comment:
In case of premature failure on `X.Close()` or `os.close(X)` the remaining pipes/fds will remain "open". Perhaps it makes sense to use `contextlib.ExitStack`.

Reply:
Is it a regression caused by my PR? Or a general remark on existing code? My PR just **moves** code, no?

Reply:
Yep, it doesn't look like a regression introduced by you (was already there).

Reply:
At least, I factorized the code so it should help to enhance the code ;-) Maybe open an issue if you want to work on that? I don't see an obvious pattern to ensure that all file descriptors are properly closed.

Reply:
Yeah, makes sense. I bet there are many other places which can benefit from `ExitStack` in a similar manner.